PolyEngine/Core/Src/pe/core/utils/Result.hpp

```diff
@@ -33,18 +52,67 @@ namespace pe::core::utils
 	{
 	public:
 		Result() = default;
-		Result(Result&) = delete;
+		Result(Result&) = default;
 		Result(Result&&) = default;
-		Result& operator=(Result&) = delete;
+		Result& operator=(Result&) = default;
 		Result& operator=(Result&&) = default;
 		Result(E errVal) : m_value(std::move(errVal)) {}
-	
-		bool IsOk() const { return m_value.has_value() == 0; }
-		bool IsErr() const { return !IsOk(); }
-		E& GetError() { return m_value.value(); }
-		const E& GetError() const { return m_value.value(); }
-	
+
+		bool isOk() const { return m_value.has_value() == false; }
+		bool isErr() const { return !isOk(); }
+
+		E& getError() { return m_value.value(); }
+		const E& getError() const { return m_value.value(); }
+
+		template <typename U>
+		Result<U, E> and(Result<U, E> res)
+		{
+			if (isOk())
+				return res;
+			else
+				return getError();
+		}
+
+		template <typename F>
+		Result<void, F> or (Result<void, F> res)
+		{
+			if (isOk())
+				return Result<void, F>();
+			else
+				return res;
+		}
+
 	private:
 		std::optional<E> m_value;
 	};
+
+	template <typename T, typename E>
+	Result<T, E> join(Result<Result<T, E>, E> res)
+	{
+		if (res.isOk())
+			return res.getValue();
+		else
+			return res.getError();
+	}
+
+	template <typename E>
+	Result<void, E> join(Result<Result<void, E>, E> res)
+	{
+		if (res.isOk())
+			return Result<void, E>();
+		else
+			return res.getError();
+	}
+
+	using aaa = void(*)();
+
+	template <typename T1, typename T2, typename E>
+	Result<T2, E> bind(Result<T1, E> res, Result<T2, E> (*func)(T1))
```

Comment:
This takes only function pointer. You casting closure (lambda with capture) to pointer is UB. Use auto

Reply:
I think auto in an function argument declaration is C++20 feature. Use std::function.
Also, this as well might be a simple method.

Reply:
Also please change T1 function argument to r-value ref and add std::move  for line 112.